{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import standard python libraries\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "import pandas as pd\n",
    "import os, subprocess\n",
    "\n",
    "# Import python package for working with cooler files and tools for analysis\n",
    "import cooler\n",
    "import cooltools.lib.plotting\n",
    "import cooltools\n",
    "\n",
    "from packaging import version\n",
    "if version.parse(cooltools.__version__) < version.parse('0.5.1'):\n",
    "    raise AssertionError(\"tutorials rely on cooltools version 0.5.1 or higher,\"+\n",
    "                         \"please check your cooltools version and update to the latest\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "from cytoolz import merge\n",
    "\n",
    "def saddleplot(\n",
    "    track,\n",
    "    saddledata,\n",
    "    n_bins,\n",
    "    vrange=None,\n",
    "    qrange=(0.0, 1.0),\n",
    "    cmap=\"coolwarm\",\n",
    "    scale=\"log\",\n",
    "    vmin=0.5,\n",
    "    vmax=2,\n",
    "    color=None,\n",
    "    title=None,\n",
    "    xlabel=None,\n",
    "    ylabel=None,\n",
    "    clabel=None,\n",
    "    fig=None,\n",
    "    fig_kws=None,\n",
    "    heatmap_kws=None,\n",
    "    margin_kws=None,\n",
    "    cbar_kws=None,\n",
    "    subplot_spec=None,\n",
    "):\n",
    "    \"\"\"\n",
    "    Generate a saddle plot.\n",
    "    Parameters\n",
    "    ----------\n",
    "    track : pd.DataFrame\n",
    "        See cooltools.digitize() for details.\n",
    "    saddledata : 2D array-like\n",
    "        Saddle matrix produced by `make_saddle`. It will include 2 flanking\n",
    "        rows/columns for outlier signal values, thus the shape should be\n",
    "        `(n+2, n+2)`.\n",
    "    cmap : str or matplotlib colormap\n",
    "        Colormap to use for plotting the saddle heatmap\n",
    "    scale : str\n",
    "        Color scaling to use for plotting the saddle heatmap: log or linear\n",
    "    vmin, vmax : float\n",
    "        Value limits for coloring the saddle heatmap\n",
    "    color : matplotlib color value\n",
    "        Face color for margin bar plots\n",
    "    fig : matplotlib Figure, optional\n",
    "        Specified figure to plot on. A new figure is created if none is\n",
    "        provided.\n",
    "    fig_kws : dict, optional\n",
    "        Passed on to `plt.Figure()`\n",
    "    heatmap_kws : dict, optional\n",
    "        Passed on to `ax.imshow()`\n",
    "    margin_kws : dict, optional\n",
    "        Passed on to `ax.bar()` and `ax.barh()`\n",
    "    cbar_kws : dict, optional\n",
    "        Passed on to `plt.colorbar()`\n",
    "    subplot_spec : GridSpec object\n",
    "        Specify a subregion of a figure to using a GridSpec.\n",
    "    Returns\n",
    "    -------\n",
    "    Dictionary of axes objects.\n",
    "    \"\"\"\n",
    "\n",
    "#     warnings.warn(\n",
    "#         \"Generating a saddleplot will be deprecated in future versions, \"\n",
    "#         + \"please see https://github.com/open2c_examples for examples on how to plot saddles.\",\n",
    "#         DeprecationWarning,\n",
    "#     )\n",
    "\n",
    "    from matplotlib.gridspec import GridSpec, GridSpecFromSubplotSpec\n",
    "    from matplotlib.colors import Normalize, LogNorm\n",
    "    from matplotlib import ticker\n",
    "    import matplotlib.pyplot as plt\n",
    "\n",
    "    class MinOneMaxFormatter(ticker.LogFormatter):\n",
    "        def set_locs(self, locs=None):\n",
    "            self._sublabels = set([vmin % 10 * 10, vmax % 10, 1])\n",
    "\n",
    "        def __call__(self, x, pos=None):\n",
    "            if x not in [vmin, 1, vmax]:\n",
    "                return \"\"\n",
    "            else:\n",
    "                return \"{x:g}\".format(x=x)\n",
    "\n",
    "    track_value_col = track.columns[3]\n",
    "    track_values = track[track_value_col].values\n",
    "\n",
    "    digitized_track, binedges = cooltools.digitize(\n",
    "        track, n_bins, vrange=vrange, qrange=qrange\n",
    "    )\n",
    "    x = digitized_track[digitized_track.columns[3]].values.astype(int).copy()\n",
    "    x = x[(x > -1) & (x < len(binedges) + 1)]\n",
    "    \n",
    "    # Old version\n",
    "    # hist = np.bincount(x, minlength=len(binedges) + 1)\n",
    "\n",
    "    groupmean = track[track.columns[3]].groupby(digitized_track[digitized_track.columns[3]]).mean()\n",
    "    \n",
    "    if qrange is not None:\n",
    "        lo, hi = qrange\n",
    "        binedges = np.linspace(lo, hi, n_bins + 1)\n",
    "    \n",
    "    # Barplot of mean values and saddledata are flanked by outlier bins\n",
    "    n = saddledata.shape[0]\n",
    "    X, Y = np.meshgrid(binedges, binedges)\n",
    "    C = saddledata\n",
    "    if (n - n_bins) == 2:\n",
    "        C = C[1:-1, 1:-1]\n",
    "        groupmean = groupmean[1:-1]\n",
    "\n",
    "    # Layout\n",
    "    if subplot_spec is not None:\n",
    "        GridSpec = partial(GridSpecFromSubplotSpec, subplot_spec=subplot_spec)\n",
    "    grid = {}\n",
    "    gs = GridSpec(\n",
    "        nrows=3,\n",
    "        ncols=3,\n",
    "        width_ratios=[0.2, 1, 0.1],\n",
    "        height_ratios=[0.2, 1, 0.1],\n",
    "        wspace=0.05,\n",
    "        hspace=0.05,\n",
    "    )\n",
    "\n",
    "    # Figure\n",
    "    if fig is None:\n",
    "        fig_kws_default = dict(figsize=(5, 5))\n",
    "        fig_kws = merge(fig_kws_default, fig_kws if fig_kws is not None else {})\n",
    "        fig = plt.figure(**fig_kws)\n",
    "\n",
    "    # Heatmap\n",
    "    if scale == \"log\":\n",
    "        norm = LogNorm(vmin=vmin, vmax=vmax)\n",
    "    elif scale == \"linear\":\n",
    "        norm = Normalize(vmin=vmin, vmax=vmax)\n",
    "    else:\n",
    "        raise ValueError(\"Only linear and log color scaling is supported\")\n",
    "\n",
    "    grid[\"ax_heatmap\"] = ax = plt.subplot(gs[4])\n",
    "    heatmap_kws_default = dict(cmap=\"coolwarm\", rasterized=True)\n",
    "    heatmap_kws = merge(\n",
    "        heatmap_kws_default, heatmap_kws if heatmap_kws is not None else {}\n",
    "    )\n",
    "    img = ax.pcolormesh(X, Y, C, norm=norm, **heatmap_kws)\n",
    "    plt.gca().yaxis.set_visible(False)\n",
    "\n",
    "    # Margins\n",
    "    margin_kws_default = dict(edgecolor=\"k\", facecolor=color, linewidth=1)\n",
    "    margin_kws = merge(margin_kws_default, margin_kws if margin_kws is not None else {})\n",
    "    # left margin hist\n",
    "    grid[\"ax_margin_y\"] = plt.subplot(gs[3], sharey=grid[\"ax_heatmap\"])\n",
    "    \n",
    "    plt.barh(\n",
    "        binedges, height=1/len(binedges), width=groupmean, align=\"edge\", **margin_kws\n",
    "    )\n",
    "    \n",
    "    plt.xlim(plt.xlim()[1], plt.xlim()[0])  # fliplr\n",
    "    plt.ylim(hi, lo)\n",
    "    plt.gca().spines[\"top\"].set_visible(False)\n",
    "    plt.gca().spines[\"bottom\"].set_visible(False)\n",
    "    plt.gca().spines[\"left\"].set_visible(False)\n",
    "    plt.gca().xaxis.set_visible(False)\n",
    "    # top margin hist\n",
    "    grid[\"ax_margin_x\"] = plt.subplot(gs[1], sharex=grid[\"ax_heatmap\"])\n",
    "    \n",
    "    plt.bar(\n",
    "        binedges, width=1/len(binedges), height=groupmean, align=\"edge\", **margin_kws\n",
    "    )\n",
    "    \n",
    "    plt.xlim(lo, hi)\n",
    "    # plt.ylim(plt.ylim())  # correct\n",
    "    plt.gca().spines[\"top\"].set_visible(False)\n",
    "    plt.gca().spines[\"right\"].set_visible(False)\n",
    "    plt.gca().spines[\"left\"].set_visible(False)\n",
    "    plt.gca().xaxis.set_visible(False)\n",
    "    plt.gca().yaxis.set_visible(False)\n",
    "\n",
    "    # Colorbar\n",
    "    grid[\"ax_cbar\"] = plt.subplot(gs[5])\n",
    "    cbar_kws_default = dict(fraction=0.8, label=clabel or \"\")\n",
    "    cbar_kws = merge(cbar_kws_default, cbar_kws if cbar_kws is not None else {})\n",
    "    if scale == \"linear\" and vmin is not None and vmax is not None:\n",
    "        grid[\"cbar\"] = cb = plt.colorbar(img, **cbar_kws)\n",
    "        # cb.set_ticks(np.arange(vmin, vmax + 0.001, 0.5))\n",
    "        # # do linspace between vmin and vmax of 5 segments and trunc to 1 decimal:\n",
    "        decimal = 10\n",
    "        nsegments = 5\n",
    "        cd_ticks = np.trunc(np.linspace(vmin, vmax, nsegments) * decimal) / decimal\n",
    "        cb.set_ticks(cd_ticks)\n",
    "    else:\n",
    "        grid[\"cbar\"] = cb = plt.colorbar(img, format=MinOneMaxFormatter(), **cbar_kws)\n",
    "        cb.ax.yaxis.set_minor_formatter(MinOneMaxFormatter())\n",
    "\n",
    "    # extra settings\n",
    "    grid[\"ax_heatmap\"].set_xlim(lo, hi)\n",
    "    grid[\"ax_heatmap\"].set_ylim(hi, lo)\n",
    "    plt.grid(False)\n",
    "    plt.axis(\"off\")\n",
    "    if title is not None:\n",
    "        grid[\"ax_margin_x\"].set_title(title)\n",
    "    if xlabel is not None:\n",
    "        grid[\"ax_heatmap\"].set_xlabel(xlabel)\n",
    "    if ylabel is not None:\n",
    "        grid[\"ax_margin_y\"].set_ylabel(ylabel)\n",
    "\n",
    "    return grid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>GC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr_1</td>\n",
       "      <td>0</td>\n",
       "      <td>5126</td>\n",
       "      <td>0.391533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr_1</td>\n",
       "      <td>5126</td>\n",
       "      <td>10252</td>\n",
       "      <td>0.424698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr_1</td>\n",
       "      <td>10252</td>\n",
       "      <td>15378</td>\n",
       "      <td>0.378268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr_1</td>\n",
       "      <td>15378</td>\n",
       "      <td>20504</td>\n",
       "      <td>0.398453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr_1</td>\n",
       "      <td>20504</td>\n",
       "      <td>25630</td>\n",
       "      <td>0.372037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40002</th>\n",
       "      <td>chr_13</td>\n",
       "      <td>11641146</td>\n",
       "      <td>11646272</td>\n",
       "      <td>0.224346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40003</th>\n",
       "      <td>chr_13</td>\n",
       "      <td>11646272</td>\n",
       "      <td>11651398</td>\n",
       "      <td>0.410066</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40004</th>\n",
       "      <td>chr_13</td>\n",
       "      <td>11651398</td>\n",
       "      <td>11656524</td>\n",
       "      <td>0.436403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40005</th>\n",
       "      <td>chr_13</td>\n",
       "      <td>11656524</td>\n",
       "      <td>11661650</td>\n",
       "      <td>0.521264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40006</th>\n",
       "      <td>chr_13</td>\n",
       "      <td>11661650</td>\n",
       "      <td>11662019</td>\n",
       "      <td>0.566396</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>40007 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom     start       end        GC\n",
       "0       chr_1         0      5126  0.391533\n",
       "1       chr_1      5126     10252  0.424698\n",
       "2       chr_1     10252     15378  0.378268\n",
       "3       chr_1     15378     20504  0.398453\n",
       "4       chr_1     20504     25630  0.372037\n",
       "...       ...       ...       ...       ...\n",
       "40002  chr_13  11641146  11646272  0.224346\n",
       "40003  chr_13  11646272  11651398  0.410066\n",
       "40004  chr_13  11651398  11656524  0.436403\n",
       "40005  chr_13  11656524  11661650  0.521264\n",
       "40006  chr_13  11661650  11662019  0.566396\n",
       "\n",
       "[40007 rows x 4 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "clr = cooler.Cooler('../cool_input/mlei_10252Cload.cool')  # 20K bins per genome of Mnemiopsis leidyi\n",
    "\n",
    "## fasta sequence is required for calculating binned profile of GC conent\n",
    "import bioframe\n",
    "bins = clr.bins()[:]\n",
    "genome = bioframe.load_fasta('../data/genome/Mlei_gDNA.fasta');\n",
    "\n",
    "## note the next command may require installing pysam\n",
    "gc_cov = bioframe.frac_gc(bins[['chrom', 'start', 'end']], genome)\n",
    "display(gc_cov)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr_1</td>\n",
       "      <td>0</td>\n",
       "      <td>11404665</td>\n",
       "      <td>chr_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr_2</td>\n",
       "      <td>0</td>\n",
       "      <td>13041538</td>\n",
       "      <td>chr_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr_3</td>\n",
       "      <td>0</td>\n",
       "      <td>15865815</td>\n",
       "      <td>chr_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr_4</td>\n",
       "      <td>0</td>\n",
       "      <td>14925576</td>\n",
       "      <td>chr_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr_5</td>\n",
       "      <td>0</td>\n",
       "      <td>17370771</td>\n",
       "      <td>chr_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr_6</td>\n",
       "      <td>0</td>\n",
       "      <td>21511028</td>\n",
       "      <td>chr_6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr_7</td>\n",
       "      <td>0</td>\n",
       "      <td>18526857</td>\n",
       "      <td>chr_7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr_8</td>\n",
       "      <td>0</td>\n",
       "      <td>20866784</td>\n",
       "      <td>chr_8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr_9</td>\n",
       "      <td>0</td>\n",
       "      <td>17080198</td>\n",
       "      <td>chr_9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr_10</td>\n",
       "      <td>0</td>\n",
       "      <td>12565900</td>\n",
       "      <td>chr_10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr_11</td>\n",
       "      <td>0</td>\n",
       "      <td>14606143</td>\n",
       "      <td>chr_11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr_12</td>\n",
       "      <td>0</td>\n",
       "      <td>15610916</td>\n",
       "      <td>chr_12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr_13</td>\n",
       "      <td>0</td>\n",
       "      <td>11662019</td>\n",
       "      <td>chr_13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom  start       end    name\n",
       "0    chr_1      0  11404665   chr_1\n",
       "1    chr_2      0  13041538   chr_2\n",
       "2    chr_3      0  15865815   chr_3\n",
       "3    chr_4      0  14925576   chr_4\n",
       "4    chr_5      0  17370771   chr_5\n",
       "5    chr_6      0  21511028   chr_6\n",
       "6    chr_7      0  18526857   chr_7\n",
       "7    chr_8      0  20866784   chr_8\n",
       "8    chr_9      0  17080198   chr_9\n",
       "9   chr_10      0  12565900  chr_10\n",
       "10  chr_11      0  14606143  chr_11\n",
       "11  chr_12      0  15610916  chr_12\n",
       "12  chr_13      0  11662019  chr_13"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "view_df = pd.DataFrame({'chrom': clr.chromnames,\n",
    "                        'start': 0,\n",
    "                        'end': clr.chromsizes.values,\n",
    "                        'name': clr.chromnames}\n",
    "                      )\n",
    "display(view_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# obtain first 3 eigenvectors\n",
    "cis_eigs = cooltools.eigs_cis(\n",
    "                        clr,\n",
    "                        gc_cov,\n",
    "                        view_df=view_df,\n",
    "                        n_eigs=3,\n",
    "                        )\n",
    "\n",
    "eigenvector_track = cis_eigs[1][['chrom','start','end', 'E1', 'E2','E3']]\n",
    "eigenvector_track.to_csv('../compartmentalization/eigenvalues/mlei_E1_3_10252bp.bed', index = False, sep = \"\\t\")\n",
    "\n",
    "# Since centromere locations are not available for the examined genomes, we manually reviewed the E1, E2, and E3 eigenvectors\n",
    "# for each chromosome to ensure that the signal accurately reflects compartmentalization patterns. Corrected eigenvalues were further\n",
    "# used to visualize the comparmentalization preferences by using saddleplots.\n",
    "\n",
    "eigenvector_track = bioframe.read_table('../compartmentalization/eigenvalues/mlei_E1_10252bp.bed')\n",
    "eigenvector_track.columns = ['chrom','start','end','E1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "cvd = cooltools.expected_cis(\n",
    "        clr=clr,\n",
    "        view_df=view_df,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>region1</th>\n",
       "      <th>region2</th>\n",
       "      <th>dist</th>\n",
       "      <th>n_valid</th>\n",
       "      <th>count.sum</th>\n",
       "      <th>balanced.sum</th>\n",
       "      <th>count.avg</th>\n",
       "      <th>balanced.avg</th>\n",
       "      <th>balanced.avg.smoothed</th>\n",
       "      <th>balanced.avg.smoothed.agg</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr_1</td>\n",
       "      <td>chr_1</td>\n",
       "      <td>0</td>\n",
       "      <td>2009</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr_1</td>\n",
       "      <td>chr_1</td>\n",
       "      <td>1</td>\n",
       "      <td>1899</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000266</td>\n",
       "      <td>0.000261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr_1</td>\n",
       "      <td>chr_1</td>\n",
       "      <td>2</td>\n",
       "      <td>1862</td>\n",
       "      <td>515782.0</td>\n",
       "      <td>47.437650</td>\n",
       "      <td>277.004296</td>\n",
       "      <td>0.025477</td>\n",
       "      <td>0.023611</td>\n",
       "      <td>0.023088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr_1</td>\n",
       "      <td>chr_1</td>\n",
       "      <td>3</td>\n",
       "      <td>1842</td>\n",
       "      <td>320855.0</td>\n",
       "      <td>30.486275</td>\n",
       "      <td>174.188382</td>\n",
       "      <td>0.016551</td>\n",
       "      <td>0.016164</td>\n",
       "      <td>0.016177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr_1</td>\n",
       "      <td>chr_1</td>\n",
       "      <td>4</td>\n",
       "      <td>1832</td>\n",
       "      <td>236735.0</td>\n",
       "      <td>22.585958</td>\n",
       "      <td>129.222162</td>\n",
       "      <td>0.012329</td>\n",
       "      <td>0.012016</td>\n",
       "      <td>0.012246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40002</th>\n",
       "      <td>chr_13</td>\n",
       "      <td>chr_13</td>\n",
       "      <td>2271</td>\n",
       "      <td>2</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.001348</td>\n",
       "      <td>4.500000</td>\n",
       "      <td>0.000674</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>0.000033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40003</th>\n",
       "      <td>chr_13</td>\n",
       "      <td>chr_13</td>\n",
       "      <td>2272</td>\n",
       "      <td>1</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.003774</td>\n",
       "      <td>8.000000</td>\n",
       "      <td>0.003774</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>0.000033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40004</th>\n",
       "      <td>chr_13</td>\n",
       "      <td>chr_13</td>\n",
       "      <td>2273</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>0.000033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40005</th>\n",
       "      <td>chr_13</td>\n",
       "      <td>chr_13</td>\n",
       "      <td>2274</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>0.000033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40006</th>\n",
       "      <td>chr_13</td>\n",
       "      <td>chr_13</td>\n",
       "      <td>2275</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>0.000033</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>40007 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      region1 region2  dist  n_valid  count.sum  balanced.sum   count.avg  \\\n",
       "0       chr_1   chr_1     0     2009        NaN           NaN         NaN   \n",
       "1       chr_1   chr_1     1     1899        NaN           NaN         NaN   \n",
       "2       chr_1   chr_1     2     1862   515782.0     47.437650  277.004296   \n",
       "3       chr_1   chr_1     3     1842   320855.0     30.486275  174.188382   \n",
       "4       chr_1   chr_1     4     1832   236735.0     22.585958  129.222162   \n",
       "...       ...     ...   ...      ...        ...           ...         ...   \n",
       "40002  chr_13  chr_13  2271        2        9.0      0.001348    4.500000   \n",
       "40003  chr_13  chr_13  2272        1        8.0      0.003774    8.000000   \n",
       "40004  chr_13  chr_13  2273        0        0.0      0.000000         NaN   \n",
       "40005  chr_13  chr_13  2274        0        0.0      0.000000         NaN   \n",
       "40006  chr_13  chr_13  2275        0        0.0      0.000000         NaN   \n",
       "\n",
       "       balanced.avg  balanced.avg.smoothed  balanced.avg.smoothed.agg  \n",
       "0               NaN                    NaN                        NaN  \n",
       "1               NaN               0.000266                   0.000261  \n",
       "2          0.025477               0.023611                   0.023088  \n",
       "3          0.016551               0.016164                   0.016177  \n",
       "4          0.012329               0.012016                   0.012246  \n",
       "...             ...                    ...                        ...  \n",
       "40002      0.000674               0.000041                   0.000033  \n",
       "40003      0.003774               0.000041                   0.000033  \n",
       "40004           NaN               0.000041                   0.000033  \n",
       "40005           NaN               0.000041                   0.000033  \n",
       "40006           NaN               0.000041                   0.000033  \n",
       "\n",
       "[40007 rows x 10 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Q_LO = 0.025 # ignore 2.5% of genomic bins with the lowest E1 values\n",
    "Q_HI = 0.975 # ignore 2.5% of genomic bins with the highest E1 values\n",
    "N_GROUPS = 38 # divide remaining 95% of the genome into 38 equisized groups, 2.5% each\n",
    "cvd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "interaction_sum, interaction_count =  cooltools.saddle(\n",
    "        clr,\n",
    "        cvd,\n",
    "        eigenvector_track,\n",
    "        'cis',\n",
    "        n_bins=N_GROUPS,\n",
    "        qrange=(Q_LO,Q_HI),\n",
    "        view_df=view_df\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[2.26268559 1.66510578 1.59627419 ... 0.65071733 0.60621437 0.50639008]\n",
      " [1.66510578 1.50229456 1.41579889 ... 0.71651192 0.68655043 0.55671721]\n",
      " [1.59627419 1.41579889 1.34961224 ... 0.73191031 0.70594432 0.6037439 ]\n",
      " ...\n",
      " [0.65071733 0.71651192 0.73191031 ... 1.1230475  1.16449523 1.26776549]\n",
      " [0.60621437 0.68655043 0.70594432 ... 1.16449523 1.23497233 1.35192556]\n",
      " [0.50639008 0.55671721 0.6037439  ... 1.26776549 1.35192556 1.52470635]]\n"
     ]
    }
   ],
   "source": [
    "value_strength_b_a = interaction_sum/interaction_count\n",
    "print(value_strength_b_a)\n",
    "np.savetxt('../data/compartmentalization/20K_bins/mlei_saddle_strength_10252bp_B-A.tsv', value_strength_b_a, delimiter='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "saddleplot(eigenvector_track,\n",
    "           interaction_sum/interaction_count,\n",
    "           N_GROUPS,\n",
    "           qrange=(Q_LO,Q_HI),\n",
    "           cbar_kws={'label':'average observed/expected contact frequency'}\n",
    "          );\n",
    "plt.savefig('../data/compartmentalization/20K_bins/Mlei_saddle_plot_10252bpRes.pdf', transparent=True, bbox_inches='tight')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cooltools_env",
   "language": "python",
   "name": "cooltools_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
